ffmpeg.c

```diff
@@ -321,109 +289,399 @@ static AVOutputFormat *get_oformat(const char *codec, char *filename){
+ *      Number of bytes written or -1 if any error happens.
  */
-struct ffmpeg *ffmpeg_open(const char *ffmpeg_video_codec, char *filename,
-                           unsigned char *y, unsigned char *u, unsigned char *v,
-                           int width, int height, int rate, int bps, int vbr, int tlapse,
-                           const struct timeval *tv1)
-{
-    AVCodecContext *c;
-    AVCodec *codec;
-    struct ffmpeg *ffmpeg;
+int ffmpeg_put_frame(struct ffmpeg *ffmpeg, AVFrame *pic, const struct timeval *tv1){
+/**
+ * Since the logic,return values and conditions changed so
+ * dramatically between versions, the encoding of the frame
+ * is 100% blocked based upon Libav/FFMpeg version
+ */
+#if (LIBAVFORMAT_VERSION_MAJOR >= 55) || ((LIBAVFORMAT_VERSION_MAJOR == 54) && (LIBAVFORMAT_VERSION_MINOR > 6))
     int retcd;
+    int got_packet_ptr;
+    AVPacket pkt;
     char errstr[128];
-    AVDictionary *opts = 0;
-
-    /*
-     * Allocate space for our ffmpeg structure. This structure contains all the
-     * codec and image information we need to generate movies.
-     */
-    ffmpeg = mymalloc(sizeof(struct ffmpeg));
+    int64_t pts_interval;
 
-    ffmpeg->tlapse  = tlapse;
 
-    /* Store codec name in ffmpeg->codec, with buffer overflow check. */
-    snprintf(ffmpeg->codec, sizeof(ffmpeg->codec), "%s", ffmpeg_video_codec);
+    av_init_packet(&pkt); /* Init static structure. */
+    if (ffmpeg->oc->oformat->flags & AVFMT_RAWPICTURE) {
+        pkt.stream_index = ffmpeg->video_st->index;
+        pkt.flags |= AV_PKT_FLAG_KEY;
+        pkt.data = (uint8_t *)pic;
+        pkt.size = sizeof(AVPicture);
+    } else {
+        pkt.data = NULL;
+        pkt.size = 0;
+        retcd = avcodec_encode_video2(AVSTREAM_CODEC_PTR(ffmpeg->video_st),
+                                        &pkt, pic, &got_packet_ptr);
+        if (retcd < 0 ){
+            av_strerror(retcd, errstr, sizeof(errstr));
+            MOTION_LOG(ERR, TYPE_ENCODER, SHOW_ERRNO, "%s: Error encoding video:%s",errstr);
+            //Packet is freed upon failure of encoding
+            return -1;
+        }
+        if (got_packet_ptr == 0){
+            //Buffered packet.  Throw special return code
+            my_packet_unref(pkt);
+            return -2;
+        }
+    }
+    if (ffmpeg->tlapse == TIMELAPSE_APPEND) {
+        retcd = timelapse_append(ffmpeg, pkt);
+    } else if (ffmpeg->tlapse == TIMELAPSE_NEW) {
+        retcd = av_write_frame(ffmpeg->oc, &pkt);
+    } else {
+        pts_interval = ((1000000L * (tv1->tv_sec - ffmpeg->start_time.tv_sec)) + tv1->tv_usec - ffmpeg->start_time.tv_usec) + 10000;
 
-    /* Allocation the output media context. */
-    ffmpeg->oc = avformat_alloc_context();
+//        MOTION_LOG(ERR, TYPE_ENCODER, NO_ERRNO, "%s: interval:%d img_sec:%d img_usec:%d strt_sec:%d strt_usec:%d "
+//                   ,pts_interval,tv1->tv_sec,tv1->tv_usec,ffmpeg->start_time.tv_sec,ffmpeg->start_time.tv_usec);
```

Comment:
Did you mean to include these lines? (Possibly they should be debug level and not commented out?)